Optimize Intersection Observer computations slightly

smfr opened the pull request:
#### 39a323e0f73673f8bf83f701cf6ee83a130d7cbe
<pre>
Optimize Intersection Observer computations slightly
<a href="https://bugs.webkit.org/show_bug.cgi?id=259630">https://bugs.webkit.org/show_bug.cgi?id=259630</a>
rdar://113083595

Reviewed by Ryosuke Niwa.

Refactor computeIntersectionState() so that we can lazily compute intersectionState.absoluteRootBounds()
only when we know we&apos;re going to fire an observation.

This requires passing the IntersectionObserverRegistration to computeIntersectionState(), which now
makes more sense as a member function, and moving the computation of thresholdIndex into this
function, so we can compute absoluteRootBounds and absoluteTargetRect lazily when we know we&apos;re going
to fire the observation.

IntersectionObservationState becomes a class-private struct since the member function returns it.
To make it clearer which rectangles have valid values in IntersectionObservationState, use
std::optional&lt;FloatRect&gt;s. Rename &quot;localRootBounds&quot; to &quot;rootBounds&quot; because the &quot;local&quot; there didn&apos;t
add any information.

Make extensive use of lambda functions in IntersectionObserver::computeIntersectionState() to
group related code and allow for early returns for clarity.

For RenderInlines, replace a call to target.boundingAbsoluteRectWithoutLayout() by code that
calls absoluteQuads() directly, since this is what boundingAbsoluteRectWithoutLayout() does for
inlines. Add a note about how inefficient this is.

There should be no behavior change here. Tested by WPT Intersection Observer tests.

* Source/WebCore/page/IntersectionObserver.cpp:
(WebCore::expandRootBoundsWithRootMargin):
(WebCore::IntersectionObserver::computeIntersectionState const):
(WebCore::IntersectionObserver::updateObservations):
(): Deleted.
(WebCore::computeIntersectionState): Deleted.
* Source/WebCore/page/IntersectionObserver.h:

Canonical link: <a href="https://commits.webkit.org/266458@main">https://commits.webkit.org/266458@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e760c938feba5c0d302a40f64832d04f676eddc2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [⏳ 🧪 style ](https://ews-build.webkit.org/#/builders/Style-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-16-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wincairo ](https://ews-build.webkit.org/#/builders/WinCairo-EWS "Waiting in queue, processing has not started yet") 
| [⏳ 🧪 bindings ](https://ews-build.webkit.org/#/builders/Bindings-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 webkitperl ](https://ews-build.webkit.org/#/builders/WebKitPerl-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-mac ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 webkitpy ](https://ews-build.webkit.org/#/builders/WebKitPy-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk1 ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 gtk-wk2 ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-16-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 services ](https://ews-build.webkit.org/#/builders/Services-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2-stress ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3369 "Built successfully and passed tests") | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-9-Build-EWS "Waiting in queue, processing has not started yet") | | [⏳ 🛠 jsc-mips ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-9-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | [⏳ 🧪 jsc-mips-tests ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/page/IntersectionObserver.cpp:
```diff
@@ -310,78 +303,133 @@ static std::optional<LayoutRect> computeClippedRectInRootContentsSpace(const Lay
+            return std::nullopt;
+
+        if (root()) {
+            auto visibleRectOptions = OptionSet<RenderObject::VisibleRectContextOption> {
```
Can't we just say `OptionSet { ~ }` instead?

smfr committed:
Optimize Intersection Observer computations slightly
https://bugs.webkit.org/show_bug.cgi?id=259630
rdar://113083595

Reviewed by Ryosuke Niwa.

Refactor computeIntersectionState() so that we can lazily compute intersectionState.absoluteRootBounds()
only when we know we're going to fire an observation.

This requires passing the IntersectionObserverRegistration to computeIntersectionState(), which now
makes more sense as a member function, and moving the computation of thresholdIndex into this
function, so we can compute absoluteRootBounds and absoluteTargetRect lazily when we know we're going
to fire the observation.

IntersectionObservationState becomes a class-private struct since the member function returns it.
To make it clearer which rectangles have valid values in IntersectionObservationState, use
std::optional<FloatRect>s. Rename "localRootBounds" to "rootBounds" because the "local" there didn't
add any information.

Make extensive use of lambda functions in IntersectionObserver::computeIntersectionState() to
group related code and allow for early returns for clarity.

For RenderInlines, replace a call to target.boundingAbsoluteRectWithoutLayout() by code that
calls absoluteQuads() directly, since this is what boundingAbsoluteRectWithoutLayout() does for
inlines. Add a note about how inefficient this is.

There should be no behavior change here. Tested by WPT Intersection Observer tests.

* Source/WebCore/page/IntersectionObserver.cpp:
(WebCore::expandRootBoundsWithRootMargin):
(WebCore::IntersectionObserver::computeIntersectionState const):
(WebCore::IntersectionObserver::updateObservations):
(): Deleted.
(WebCore::computeIntersectionState): Deleted.
* Source/WebCore/page/IntersectionObserver.h:

Canonical link: https://commits.webkit.org/266458@main
